sparse_ho/tests/test_models.py

```diff
@@ -0,0 +1,282 @@
+    dict_vals_cvxpy[model, 'MSE'] = val_cvxpy
+    grad_cvxpy *= np.exp(dict_log_alpha[model])
+    dict_grads_cvxpy[model, 'MSE'] = grad_cvxpy
+
+
+val_cvxpy, grad_cvxpy = lasso_sure_cvxpy(
+    X, y, np.exp(dict_log_alpha["lasso"]), sigma_star)
+grad_cvxpy *= np.exp(dict_log_alpha["lasso"])
+dict_vals_cvxpy["lasso", "SURE"] = val_cvxpy
+dict_grads_cvxpy["lasso", "SURE"] = grad_cvxpy
+
+# log alpha to be tested by checkgrad
+dict_list_log_alphas = {}
+dict_list_log_alphas["lasso"] = np.log(
+    np.geomspace(alpha_max/2, alpha_max/5, num=5))
+dict_list_log_alphas["wLasso"] = [
+    log_alpha * np.ones(n_features) for log_alpha in
+    dict_list_log_alphas["lasso"]]
+dict_list_log_alphas["logreg"] = np.log(
+    np.geomspace(alpha_max/5, alpha_max/40, num=5))
+dict_list_log_alphas["enet"] = [np.array(i) for i in itertools.product(
+    dict_list_log_alphas["lasso"], dict_list_log_alphas["lasso"])]
+
+
+def get_v(mask, dense):
+    return 2 * (X[np.ix_(idx_val, mask)].T @ (
+        X[np.ix_(idx_val, mask)] @ dense - y[idx_val])) / len(idx_val)
+
+
+@pytest.mark.parametrize('key', list(models.keys()))
+def test_beta_jac(key):
+    """Tests that algorithms computing the Jacobian return the same Jacobian"""
+    supp1, dense1, jac1 = get_beta_jac_iterdiff(
+        X, y, dict_log_alpha[key], tol=tol, model=models[key])
+    supp2, dense2, jac2 = get_beta_jac_fast_iterdiff(
+        X, y, dict_log_alpha[key], tol=tol, model=models[key], tol_jac=tol)
+    supp3, dense3, jac3 = get_beta_jac_iterdiff(
+        X_s, y, dict_log_alpha[key], tol=tol,
+        model=models[key])
+    supp4, dense4, jac4 = get_beta_jac_fast_iterdiff(
+        X_s, y, dict_log_alpha[key],
+        tol=tol, model=models[key], tol_jac=tol)
+
+    assert np.all(supp1 == supp2)
+    assert np.allclose(dense1, dense2)
+    assert np.allclose(jac1, jac2, atol=1e-6)
+
+    assert np.all(supp2 == supp3)
+    assert np.allclose(dense2, dense3)
+    assert np.allclose(jac2, jac3, atol=1e-6)
+
+    assert np.all(supp3 == supp4)
+    assert np.allclose(dense3, dense4)
+    assert np.allclose(jac3, jac4, atol=1e-6)
+
+    get_beta_jac_t_v_implicit(
+        X, y, dict_log_alpha[key], get_v, model=models[key])
+
+
+@pytest.mark.parametrize('key', list(custom_models.keys()))
+def test_beta_jac_custom(key):
+    """Check that using sk or celer yields the same solution as sparse ho"""
+    supp, dense, jac = get_beta_jac_fast_iterdiff(
+        X_s, y, dict_log_alpha[key],
+        tol=tol, model=models[key], tol_jac=tol)
+    supp_custom, dense_custom, jac_custom = get_beta_jac_fast_iterdiff(
+        X_s, y, dict_log_alpha[key],
+        tol=tol, model=custom_models[key], tol_jac=tol)
+    assert np.all(supp == supp_custom)
+    assert np.allclose(dense, dense_custom)
+    assert np.allclose(jac, jac_custom)
+
+
+list_model_crit = [
+    ('lasso', 'MSE'),
+    ('enet', 'MSE'),
+    ('wLasso', 'MSE'),
+    ('lasso', 'SURE'),
+    # ('enet', 'SURE'),
+    # ('wLasso', 'SURE'),
```

Comment:
can you write a comment on why you have these 2 commented for next one who reads?